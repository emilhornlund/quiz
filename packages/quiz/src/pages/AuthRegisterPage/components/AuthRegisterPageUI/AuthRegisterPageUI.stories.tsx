import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-remix-react-router'

import AuthRegisterPageUI from './AuthRegisterPageUI'

const meta = {
  title: 'Pages/AuthRegisterPage',
  component: AuthRegisterPageUI,
  decorators: [withRouter],
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof AuthRegisterPageUI>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {
    loading: false,
    onSubmit: () => undefined,
  },
} satisfies Story
