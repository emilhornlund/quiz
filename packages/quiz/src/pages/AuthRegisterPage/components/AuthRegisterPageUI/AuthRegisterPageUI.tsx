import {
  CreateUserRequestDto,
  EMAIL_MAX_LENGTH,
  EMAIL_MIN_LENGTH,
  EMAIL_REGEX,
  FAMILY_NAME_MAX_LENGTH,
  FAMILY_NAME_MIN_LENGTH,
  FAMILY_NAME_REGEX,
  GIVEN_NAME_MAX_LENGTH,
  GIVEN_NAME_MIN_LENGTH,
  GIVEN_NAME_REGEX,
  PASSWORD_MAX_LENGTH,
  PASSWORD_MIN_LENGTH,
  PASSWORD_REGEX,
} from '@quiz/common'
import React, { FC, FormEvent, useMemo, useState } from 'react'
import { Link } from 'react-router-dom'

import {
  IconButtonArrowRight,
  Page,
  TextField,
  Typography,
} from '../../../../components'
import NicknameTextField from '../../../../components/NicknameTextField'

import styles from './AuthRegisterPageUI.module.scss'
import { getMessage, getTitle } from './helpers.ts'

export type CreateUserFormFields = CreateUserRequestDto

type CreateUserValidFormFields = {
  [key in keyof CreateUserFormFields]: boolean
}

export interface AuthRegisterPageUIProps {
  loading: boolean
  onSubmit: (values: CreateUserFormFields) => void
}

const AuthRegisterPageUI: FC<AuthRegisterPageUIProps> = ({
  loading,
  onSubmit,
}) => {
  const title = useMemo(getTitle, [])
  const message = useMemo(getMessage, [])

  const [formFields, setFormFields] = useState<CreateUserFormFields>({
    email: '',
    password: '',
    givenName: '',
    familyName: '',
    defaultNickname: '',
  })

  const handleChangeFormField = <K extends keyof CreateUserRequestDto>(
    key: K,
    value: CreateUserFormFields[K],
  ) => {
    setFormFields({ ...formFields, [key]: value })
  }

  const [validFormFields, setValidFormFields] =
    useState<CreateUserValidFormFields>({
      email: false,
      password: false,
      givenName: false,
      familyName: false,
      defaultNickname: false,
    })

  const isFormValid = useMemo(
    () => Object.values(validFormFields).every((valid) => !!valid),
    [validFormFields],
  )

  const handleChangeValidFormField = <
    K extends keyof CreateUserValidFormFields,
  >(
    key: K,
    valid: boolean,
  ) => {
    if (validFormFields[key] !== valid) {
      setValidFormFields((prevState) => {
        return { ...prevState, [key]: valid }
      })
    }
  }

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault()
    onSubmit({
      ...formFields,
      givenName: formFields.givenName || undefined,
      familyName: formFields.familyName || undefined,
      defaultNickname: formFields.defaultNickname,
    })
  }

  return (
    <Page>
      <Typography variant="title" size="medium">
        {title}
      </Typography>
      <Typography variant="text" size="small">
        {message}
      </Typography>
      <form className={styles.createUserForm} onSubmit={handleSubmit}>
        <TextField
          id="email"
          type="text"
          placeholder="Email"
          value={formFields.email}
          minLength={EMAIL_MIN_LENGTH}
          maxLength={EMAIL_MAX_LENGTH}
          regex={EMAIL_REGEX}
          disabled={loading}
          onChange={(value) => handleChangeFormField('email', value as string)}
          onValid={(valid) => handleChangeValidFormField('email', valid)}
          required
        />
        <TextField
          id="password"
          type="password"
          placeholder="Password"
          value={formFields.password}
          minLength={PASSWORD_MIN_LENGTH}
          maxLength={PASSWORD_MAX_LENGTH}
          regex={{
            value: PASSWORD_REGEX,
            message:
              'Must include at least ≥2 uppercase, ≥2 lowercase, ≥2 digits, ≥2 symbols.',
          }}
          disabled={loading}
          onChange={(value) =>
            handleChangeFormField('password', value as string)
          }
          onValid={(valid) => handleChangeValidFormField('password', valid)}
          required
        />
        <TextField
          id="givenName"
          type="text"
          placeholder="Given Name"
          value={formFields.givenName}
          minLength={GIVEN_NAME_MIN_LENGTH}
          maxLength={GIVEN_NAME_MAX_LENGTH}
          regex={GIVEN_NAME_REGEX}
          disabled={loading}
          onChange={(value) =>
            handleChangeFormField('givenName', value as string)
          }
          onValid={(valid) => handleChangeValidFormField('givenName', valid)}
        />
        <TextField
          id="familyName"
          type="text"
          placeholder="Family Name"
          value={formFields.familyName}
          minLength={FAMILY_NAME_MIN_LENGTH}
          maxLength={FAMILY_NAME_MAX_LENGTH}
          regex={FAMILY_NAME_REGEX}
          disabled={loading}
          onChange={(value) =>
            handleChangeFormField('familyName', value as string)
          }
          onValid={(valid) => handleChangeValidFormField('familyName', valid)}
        />
        <NicknameTextField
          value={formFields.defaultNickname}
          disabled={loading}
          onChange={(value) => handleChangeFormField('defaultNickname', value)}
          onValid={(valid) =>
            handleChangeValidFormField('defaultNickname', valid)
          }
        />
        <IconButtonArrowRight
          id="join"
          type="submit"
          kind="call-to-action"
          value="Let's go!"
          loading={loading}
          disabled={!isFormValid || loading}
        />
      </form>
      <Link to={'/auth/login'}>
        <Typography variant="link" size="small">
          Got an account? Flash your credentials and come on in!
        </Typography>
      </Link>
    </Page>
  )
}

export default AuthRegisterPageUI
