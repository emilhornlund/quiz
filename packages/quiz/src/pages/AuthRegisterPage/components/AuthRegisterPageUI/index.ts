export type {
  AuthRegisterPageUIProps,
  CreateUserFormFields,
} from './AuthRegisterPageUI'
export { default } from './AuthRegisterPageUI'
