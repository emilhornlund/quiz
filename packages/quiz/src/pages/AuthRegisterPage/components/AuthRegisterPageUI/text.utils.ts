export const TITLES = [
  'Join the quiz zone!',
  'Welcome, newcomer!',
  'Create your quiz identity',
  'Start your trivia journey',
  'Ready to play with the pros?',
  'Become a Quizmaster!',
  'Let’s get you set up!',
  'A brainy adventure awaits!',
  'Unlock your quiz future',
  'New here? Let’s fix that!',
]

export const MESSAGES = [
  'Create an account to start building quizzes, tracking your progress, and challenging others to brainy showdowns.',
  'Join a growing world of trivia lovers and unlock access to all the quizzes, stats, and features you never knew you needed.',
  'Sign up and claim your spot on the leaderboard — your quiz legacy starts here.',
  'Register now to craft quizzes, answer questions, and maybe even become a legend in the trivia universe.',
  'It all begins with a name, a password, and a spark of curiosity. Let’s get you signed up!',
  'Joining only takes a moment, but your quiz journey could last a lifetime (or at least until the next leaderboard reset).',
  'Create your account and dive headfirst into a world of questions, rankings, and brain-flexing fun.',
  'Signing up gives you access to personalized quiz history, stats, and the chance to shine in every round.',
  'Set up your profile, choose your nickname, and let the quiz games begin — well, almost.',
]

const getRandomString = (strings: string[]) => {
  const randomIndex = Math.floor(Math.random() * strings.length)
  return strings[randomIndex]
}

export const getTitle = () => getRandomString(TITLES)

export const getMessage = () => getRandomString(MESSAGES)
