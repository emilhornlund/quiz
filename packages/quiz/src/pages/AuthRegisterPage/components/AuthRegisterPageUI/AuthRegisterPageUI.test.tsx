import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { describe, expect, it, vi } from 'vitest'

import AuthRegisterPageUI from './AuthRegisterPageUI'

vi.mock('./text.utils.ts', () => ({
  getTitle: () => 'Create your account',
  getMessage: () => 'Join the party — make an account in seconds.',
}))

describe('AuthRegisterPageUI', () => {
  it('renders default state', () => {
    const { container } = render(
      <MemoryRouter>
        <AuthRegisterPageUI loading={false} onSubmit={() => undefined} />
      </MemoryRouter>,
    )
    expect(screen.getByText('Create your account')).toBeInTheDocument()
    expect(
      screen.getByText('Join the party — make an account in seconds.'),
    ).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders loading state', () => {
    const { container } = render(
      <MemoryRouter>
        <AuthRegisterPageUI loading={true} onSubmit={() => undefined} />
      </MemoryRouter>,
    )
    // Submit should be disabled while loading
    const submit = container.querySelector('#join') as HTMLButtonElement
    expect(submit).toBeDisabled()
    expect(container).toMatchSnapshot()
  })

  it('keeps submit disabled until all fields are valid, then submits values', () => {
    const onSubmit = vi.fn()
    const { container } = render(
      <MemoryRouter>
        <AuthRegisterPageUI loading={false} onSubmit={onSubmit} />
      </MemoryRouter>,
    )

    const submit = container.querySelector('#join') as HTMLButtonElement
    expect(submit).toBeDisabled()

    // Fill required/validated fields with values that satisfy regex/lengths
    fireEvent.change(screen.getByPlaceholderText('Email'), {
      target: { value: 'user@example.com' },
    })
    fireEvent.change(screen.getByPlaceholderText('Password'), {
      target: { value: 'PPss11!!' }, // ≥2 upper, ≥2 lower, ≥2 digits, ≥2 symbols
    })
    fireEvent.change(screen.getByPlaceholderText('Given Name'), {
      target: { value: 'Alice' },
    })
    fireEvent.change(screen.getByPlaceholderText('Family Name'), {
      target: { value: 'Smith' },
    })

    // NicknameTextField doesn't expose a placeholder; select the remaining text input
    const textInputs = Array.from(
      container.querySelectorAll('input[type="text"]'),
    ) as HTMLInputElement[]
    const nicknameInput = textInputs.find(
      (el) =>
        el.id !== 'email' && el.id !== 'givenName' && el.id !== 'familyName',
    ) as HTMLInputElement
    fireEvent.change(nicknameInput, { target: { value: 'FrostyBear' } })

    expect(submit).not.toBeDisabled()

    fireEvent.click(submit)

    expect(onSubmit).toHaveBeenCalledTimes(1)
    expect(onSubmit).toHaveBeenCalledWith({
      email: 'user@example.com',
      password: 'PPss11!!',
      givenName: 'Alice',
      familyName: 'Smith',
      defaultNickname: 'FrostyBear',
    })
  })
})
