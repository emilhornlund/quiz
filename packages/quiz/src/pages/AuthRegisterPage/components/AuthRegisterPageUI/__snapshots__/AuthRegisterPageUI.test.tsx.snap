// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthRegisterPageUI > renders default state 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content startAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="card kindCallToAction sizeLarge center"
          >
            <button
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-xmark"
                data-icon="xmark"
                data-prefix="fas"
                role="img"
                viewBox="0 0 384 512"
              >
                <path
                  d="M55.1 73.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L147.2 256 9.9 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192.5 301.3 329.9 438.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.8 256 375.1 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192.5 210.7 55.1 73.4z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="content"
            >
              <span>
                Hey quiz champion! Did your brainy adventures happen on
                 
                <a
                  href="https://quiz.emilhornlund.com"
                >
                  quiz.emilhornlund.com
                </a>
                ? No worries— we can save all your quiz creations and epic game stats! Just pop back to
                 
                <a
                  href="https://quiz.emilhornlund.com"
                >
                  quiz.emilhornlund.com
                </a>
                 one last time, and we’ll magically teleport everything over to
                 
                <b>
                  <i>
                    klurigo.com
                  </i>
                </b>
                . Ready to blast off your data?
              </span>
            </div>
          </div>
          <div
            class="typography title medium"
          >
            Create your account
          </div>
          <div
            class="typography text small"
          >
            Join the party — make an account in seconds.
          </div>
          <form
            class="createUserForm"
          >
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary"
              >
                <input
                  class="textfield"
                  data-testid="test-email-textfield"
                  id="email"
                  maxlength="128"
                  minlength="6"
                  name="email"
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary"
              >
                <input
                  class="textfield"
                  data-testid="test-password-textfield"
                  id="password"
                  maxlength="128"
                  minlength="8"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary"
              >
                <input
                  class="textfield"
                  data-testid="test-givenName-textfield"
                  id="givenName"
                  maxlength="64"
                  minlength="1"
                  name="givenName"
                  placeholder="Given Name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary"
              >
                <input
                  class="textfield"
                  data-testid="test-familyName-textfield"
                  id="familyName"
                  maxlength="64"
                  minlength="1"
                  name="familyName"
                  placeholder="Family Name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="nicknameTextFieldContainer"
            >
              <div
                class="inputContainer"
              >
                <div
                  class="textFieldInputContainer textFieldInputKindPrimary"
                >
                  <input
                    class="textfield"
                    data-testid="test-default-nickname-textfield-textfield"
                    id="default-nickname-textfield"
                    maxlength="20"
                    minlength="2"
                    name="default-nickname-textfield"
                    placeholder="Default Nickname"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="buttonInputContainer buttonInputKindPrimary"
              >
                <button
                  data-testid="test-shuffle-nickname-button-button"
                  id="shuffle-nickname-button"
                  name="shuffle-nickname-button"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-arrows-rotate"
                    data-icon="arrows-rotate"
                    data-prefix="fas"
                    role="img"
                    viewBox="0 0 512 512"
                  >
                    <path
                      d="M65.9 228.5c13.3-93 93.4-164.5 190.1-164.5 53 0 101 21.5 135.8 56.2 .2 .2 .4 .4 .6 .6l7.6 7.2-47.9 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-128c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 53.4-11.3-10.7C390.5 28.6 326.5 0 256 0 127 0 20.3 95.4 2.6 219.5 .1 237 12.2 253.2 29.7 255.7s33.7-9.7 36.2-27.1zm443.5 64c2.5-17.5-9.7-33.7-27.1-36.2s-33.7 9.7-36.2 27.1c-13.3 93-93.4 164.5-190.1 164.5-53 0-101-21.5-135.8-56.2-.2-.2-.4-.4-.6-.6l-7.6-7.2 47.9 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 320c-8.5 0-16.7 3.4-22.7 9.5S-.1 343.7 0 352.3l1 127c.1 17.7 14.6 31.9 32.3 31.7S65.2 496.4 65 478.7l-.4-51.5 10.7 10.1c46.3 46.1 110.2 74.7 180.7 74.7 129 0 235.7-95.4 253.4-219.5z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="buttonInputContainer buttonInputKindCallToAction"
            >
              <button
                data-testid="test-join-button"
                disabled=""
                id="join"
                name="join"
                type="submit"
              >
                <span>
                  Let's go!
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-arrow-right"
                  data-icon="arrow-right"
                  data-prefix="fas"
                  role="img"
                  viewBox="0 0 512 512"
                >
                  <path
                    d="M502.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l370.7 0-105.4 105.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </form>
          <a
            data-discover="true"
            href="/auth/login"
          >
            <div
              class="typography link small"
            >
              Got an account? Flash your credentials and come on in!
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AuthRegisterPageUI > renders loading state 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content startAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="card kindCallToAction sizeLarge center"
          >
            <button
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-xmark"
                data-icon="xmark"
                data-prefix="fas"
                role="img"
                viewBox="0 0 384 512"
              >
                <path
                  d="M55.1 73.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L147.2 256 9.9 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192.5 301.3 329.9 438.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.8 256 375.1 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192.5 210.7 55.1 73.4z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="content"
            >
              <span>
                Hey quiz champion! Did your brainy adventures happen on
                 
                <a
                  href="https://quiz.emilhornlund.com"
                >
                  quiz.emilhornlund.com
                </a>
                ? No worries— we can save all your quiz creations and epic game stats! Just pop back to
                 
                <a
                  href="https://quiz.emilhornlund.com"
                >
                  quiz.emilhornlund.com
                </a>
                 one last time, and we’ll magically teleport everything over to
                 
                <b>
                  <i>
                    klurigo.com
                  </i>
                </b>
                . Ready to blast off your data?
              </span>
            </div>
          </div>
          <div
            class="typography title medium"
          >
            Create your account
          </div>
          <div
            class="typography text small"
          >
            Join the party — make an account in seconds.
          </div>
          <form
            class="createUserForm"
          >
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary disabled"
              >
                <input
                  class="textfield"
                  data-testid="test-email-textfield"
                  disabled=""
                  id="email"
                  maxlength="128"
                  minlength="6"
                  name="email"
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary disabled"
              >
                <input
                  class="textfield"
                  data-testid="test-password-textfield"
                  disabled=""
                  id="password"
                  maxlength="128"
                  minlength="8"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary disabled"
              >
                <input
                  class="textfield"
                  data-testid="test-givenName-textfield"
                  disabled=""
                  id="givenName"
                  maxlength="64"
                  minlength="1"
                  name="givenName"
                  placeholder="Given Name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="inputContainer"
            >
              <div
                class="textFieldInputContainer textFieldInputKindPrimary disabled"
              >
                <input
                  class="textfield"
                  data-testid="test-familyName-textfield"
                  disabled=""
                  id="familyName"
                  maxlength="64"
                  minlength="1"
                  name="familyName"
                  placeholder="Family Name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="nicknameTextFieldContainer"
            >
              <div
                class="inputContainer"
              >
                <div
                  class="textFieldInputContainer textFieldInputKindPrimary disabled"
                >
                  <input
                    class="textfield"
                    data-testid="test-default-nickname-textfield-textfield"
                    disabled=""
                    id="default-nickname-textfield"
                    maxlength="20"
                    minlength="2"
                    name="default-nickname-textfield"
                    placeholder="Default Nickname"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="buttonInputContainer buttonInputKindPrimary"
              >
                <button
                  data-testid="test-shuffle-nickname-button-button"
                  disabled=""
                  id="shuffle-nickname-button"
                  name="shuffle-nickname-button"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-arrows-rotate"
                    data-icon="arrows-rotate"
                    data-prefix="fas"
                    role="img"
                    viewBox="0 0 512 512"
                  >
                    <path
                      d="M65.9 228.5c13.3-93 93.4-164.5 190.1-164.5 53 0 101 21.5 135.8 56.2 .2 .2 .4 .4 .6 .6l7.6 7.2-47.9 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-128c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 53.4-11.3-10.7C390.5 28.6 326.5 0 256 0 127 0 20.3 95.4 2.6 219.5 .1 237 12.2 253.2 29.7 255.7s33.7-9.7 36.2-27.1zm443.5 64c2.5-17.5-9.7-33.7-27.1-36.2s-33.7 9.7-36.2 27.1c-13.3 93-93.4 164.5-190.1 164.5-53 0-101-21.5-135.8-56.2-.2-.2-.4-.4-.6-.6l-7.6-7.2 47.9 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 320c-8.5 0-16.7 3.4-22.7 9.5S-.1 343.7 0 352.3l1 127c.1 17.7 14.6 31.9 32.3 31.7S65.2 496.4 65 478.7l-.4-51.5 10.7 10.1c46.3 46.1 110.2 74.7 180.7 74.7 129 0 235.7-95.4 253.4-219.5z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="buttonInputContainer buttonInputKindCallToAction"
            >
              <button
                data-testid="test-join-button"
                disabled=""
                id="join"
                name="join"
                type="submit"
              >
                <div
                  class="loadingSpinner"
                >
                  <div />
                  <div />
                  <div />
                </div>
              </button>
            </div>
          </form>
          <a
            data-discover="true"
            href="/auth/login"
          >
            <div
              class="typography link small"
            >
              Got an account? Flash your credentials and come on in!
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
